### PR TITLE
 Instrument the Ruler Notifier

### LINKED ISSUE #2193: Instrument Ruler Notifier
One of the largest gaps in monitoring Cortex is the lack of instrumentation in Ruler Notifier code path. Currently no metrics registerer is passes to the Prometheus notifier leaving a gap in visibility as to whether alerts in the ruler are actually be sent to the target Alertmanager.

### PR DESCRIPTION
**Pull request by gotjosh**
**What this PR does**:

The ruler notifier is not instrumented at the moment. This commit wraps
around the Ruler registry with the userID as a label and `cortex_` as a
prefix and passes it to the notifier options.

The result? We have metrics.

**Which issue(s) this PR fixes**:
Fixes #2193

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ruler/ruler_test.go [hunk 1]
```diff
@@ -6,13 +6,15 @@ import (
...
 	"os"
+	"strings"
 	"sync"
 	"testing"
 	"time"
 
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
 	"github.com/prometheus/client_golang/prometheus"
+	promTU "github.com/prometheus/client_golang/prometheus/testutil"
```
Comment (cstyan) [inline pkg/ruler/ruler_test.go:17]: minor nit but when we import both testutil packages in prometheus projects we name this one `prom_testutil "github.com/prometheus/client_golang/prometheus/testutil"`

**Comment by gouthamve** [comment]
LGTM pending rebase. Slightly concerning but should be okay. `prometheus_notifications_latency_seconds` is a summary adding 5 metrics per AM. And given people typically run 3 AMs, it will be an additional 15 metrics per user. I think it should be fine. If its not, we can revert later.

File: pkg/ruler/ruler_test.go (hunk 1 above)
Comment (gotjosh) [reply pkg/ruler/ruler_test.go]: Good catch.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,14 @@
...
 ## master / unreleased
 
+* [ENHANCEMENT Ruler: Added the following metrics: #2786
+  * `cortex_prometheus_notifications_latency_seconds`
+  * `cortex_prometheus_notifications_errors_total`
+  * `cortex_prometheus_notifications_sent_total`
+  * `cortex_prometheus_notifications_dropped_total`
+  * `cortex_prometheus_notifications_queue_length`
+  * `cortex_prometheus_notifications_queue_capacity`
+  * `cortex_prometheus_notifications_prometheus_notifications_alertmanagers_discovered`
```
Comment (pracucci) [inline CHANGELOG.md:12]: The metric is registered like this:
```
alertmanagersDiscovered: prometheus.NewGaugeFunc(prometheus.GaugeOpts{
			Name: "prometheus_notifications_alertmanagers_discovered",
			Help: "The number of alertmanagers discovered and active.",
		}, alertmanagersDiscovered),
```

So we should fix this.

```suggestion
  * `cortex_prometheus_notifications_alertmanagers_discovered`
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,14 @@
 
 ## master / unreleased
 
+* [ENHANCEMENT Ruler: Added the following metrics: #2786
```
Comment (pracucci) [inline CHANGELOG.md:5]: CHANGELOG entries are grouped by type. Could you move this entry below, along with the other `[ENHANCEMENT]` entries, please?

File: CHANGELOG.md (hunk 2 above)
Comment (gotjosh) [reply CHANGELOG.md]: That last rebase really messed me up 😅

**Comment by gotjosh** [comment]
I can see that the changelog is having conflicts again, just let me know if we're ready to merge and I'll fix it 😄

**Comment by pracucci** [comment]
@gotjosh I took care of CHANGELOG.

**Comment by gotjosh** [comment]
Thanks @pracucci !

**Pull request merged**